src/App.tsx

```diff
@@ -1,24 +1,197 @@
+    }, 3000);
+  };
+
+  useEffect(() => {
+    showErrorMessage();
+  }, [errorMessage]);
+
+  useEffect(() => {
+    setErrorMessage('');
+
+    dataOperations
+      .getTodos(USER_ID)
+      .then(setTodos)
+      .catch((error) => {
+        setErrorMessage('Unable to load todos');
+        setTimeout(() => {
+          setErrorMessage('');
+        }, 3000);
+        throw error;
+      });
+  }, []);
+
+  const filteredTodos = useMemo(() => {
+    return filterTodos(todos, filter);
+  }, [todos, filter]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const addTodo = ({ userId, title, completed }: Todo) => {
+    setErrorMessage('');
+
+    const temporaryTodo = {
+      userId,
+      title,
+      completed,
+      id: 0,
+    };
+
+    setTempTodo(temporaryTodo);
+    setIsTodoLoading(temporaryTodo);
+    setIsInputDisabled(true);
+
+    return dataOperations
+      .addTodoToServer({ userId, title, completed })
+      .then((newTodo) => {
+        setTodos((currentTodos) => [...currentTodos, newTodo]);
+      })
+      .catch((error) => {
+        setTodos(prev => prev.filter(todo => todo.id !== temporaryTodo.id));
+        setErrorMessage('Unable to add a todo');
+        setTimeout(() => {
+          setErrorMessage('');
+        }, 3000);
+        throw error;
+      })
+      .finally(() => {
+        setTempTodo(null);
+        setIsInputDisabled(false);
+        setIsTodoLoading(null);
+      });
+  };
+
+  const deleteTodo = (todoId: number) => {
+    setErrorMessage('');
+
+    return dataOperations.deleteTodoOnServer(todoId)
+      .then(() => {
+        setTodos((currentTodo) => currentTodo.filter((todo) => todo.id !== todoId));
+      })
+      .catch((error) => {
+        setTodos(todos);
+        setErrorMessage('Unable to delete a todo');
+        setTimeout(() => {
+          setErrorMessage('');
+        }, 3000);
+        throw error;
+      });
+  };
+
+  const onCheckedToggle = (todoId: number) => {
+    setTodos((currentTodo) => {
+      return currentTodo.map((todo) => {
+        if (todo.id === todoId) {
+          return {
+            ...todo,
+            completed: !todo.completed,
+          };
+        }
+
+        return todo;
+      });
+    });
+  };
+
+  const handleFilter = (newFilter: Filter) => {
+    setFilter(newFilter);
+  };
+
+  const count = todos.filter((todo) => todo.completed !== true).length;
+
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-loading-todos#react-todo-app-load-todos">React Todo App - Load Todos</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          {todos.length !== 0 && (
```

Comment:
```suggestion
          {!!todos.length && (
```
**PLS DO THE SAME FOR ALL SUCH CASES IN THE PROJECT**